src/tempfile.rs

```diff
@@ -0,0 +1,167 @@
+// Copyright 2017 The Chromium OS Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE-BSD-3-Clause file.
+//
+// SPDX-License-Identifier: BSD-3-Clause
+
+//! Create temporary files.
+//!
+//! The temporary files will be created with a name which is available as well
+//! as having an exposed `fs::File` for reading/writing.
+//!
+//! The file will be removed when the object goes out of scope.
+//!
+//! # Examples
+//!
+//! ```
+//! use std::io::Write;
+//! use std::path::Path;
+//! use std::path::PathBuf;
+//! use vmm_sys_util::tempfile::TempFile;
+//! let t = TempFile::new("/tmp/testfile").unwrap();
+//! let mut f = t.as_file();
+//! f.write_all(b"hello world").unwrap();
+//! f.sync_all().unwrap();
+
+use std::ffi::{CString, OsStr};
+use std::fs;
+use std::fs::File;
+use std::os::unix::ffi::{OsStrExt, OsStringExt};
+use std::os::unix::io::FromRawFd;
+use std::path::{Path, PathBuf};
+
+use libc;
+
+use crate::errno::{errno_result, Result, Error};
+
+/// Create and remove a temporary file.  The file will be maintained for the
+/// lifetime of the `TempFile` object.
+pub struct TempFile {
+    path: PathBuf,
+    file: File,
+    removed: bool,
+}
+
+
+impl TempFile {
+
+    /// Creates the TempFile
+    ///
+    /// # Arguments
+    ///
+    /// `prefix`: The path and filename where to creat the temporary file. Six
+    /// random alphanumeric characters will be added to the end of this to form
+    /// the filename.
+    pub fn new<P: AsRef<OsStr>>(prefix: P) -> Result<TempFile> {
+        let mut os_fname = prefix.as_ref().to_os_string();
+        os_fname.push("XXXXXX");
+
+        let raw_fname = CString::new(os_fname.into_vec()).unwrap().into_raw();
+
+        // Safe because I'm sure `raw_fname` is a valid CString.
+        let fd = unsafe { libc::mkstemp(raw_fname) };
+
+        // The `raw_fname` is the same as `os_fname` (which is `prefix` with X's
+        // appended). The X's are documented as "six characters" which seem to
+        // always be alphanumerics so appears safe.
+        let c_tempname = unsafe { CString::from_raw(raw_fname) };
+        let os_tempname = OsStr::from_bytes(c_tempname.as_bytes());
+
+        if fd == -1 {
+            return errno_result();
+        }
+
+        // Safe because we checked `fd != -1` above and we uniquely own the file
+        // descriptor. This `fd` will be freed etc when `File` and thus
+        // `TempFile` goes out of scope.
+        let file = unsafe { File::from_raw_fd(fd) };
+
+        Ok(TempFile {
+            path: PathBuf::from(os_tempname),
+            file,
+            removed: false,
+        })
+    }
+
+    /// Removes the temporary file.  Calling this is optional as dropping a
+    /// `TempFile` object will also remove the file.  Calling remove explicitly
+    /// allows for better error handling.
+    pub fn remove(&mut self) -> Result<()> {
+        if self.removed {
```

Comment:
A discussion about how to approach double remove also happened on this PR: https://github.com/rust-vmm/vmm-sys-util/pull/35#discussion_r317231334. I think it's better to return the Error here in case remove is called twice because that's how the `remove_file` function works as well. When called on a file that does not exist, it will return an error. I think we should just propagate that error instead of keeping an internal state with `self.removed`.

CC: @jiangliu

Reply:
I'll wait for that discussion to be resolved.

Reply:
Looks like we chose to only write a comment that an error is returned in case the function is called twice and not keep the internal state. The tempdir PR was merged: https://github.com/rust-vmm/vmm-sys-util/pull/35

Reply:
@forbesmyester once you remove the self.remove field and just return the error from `fs::remove_file` we are ready to merge this PR.

Reply:
I only have one branch in my code (that I don't think I can reliably execute). Any thoughts how I can get the 0.1% test coverage back?

Reply:
I just ran the coverage test locally on your branch + upstream changes and it looks like the coverage is increased from to 74.9. I also looked at the covered code and it seems good. Just do a rebase and add the appropriate coverage in coverage_config.json. We can merge this PR next.

Reply:
Thanks